Add beginnings of torch::stable::accelerator

mikaylagawarecki opened the pull request:
Adds 
- `torch::stable::accelerator::DeviceGuard`: `std::unique_ptr` to `DeviceGuardOpauqe` mostly copied from the below (but made generic)

   https://github.com/pytorch/pytorch/blob/50eac811a68e63e96ad56c11c983bfe298a0bb8a/torch/csrc/inductor/aoti_runtime/utils_cuda.h#L30-L46
    - constructor `DeviceGuard(DeviceIndex)` (**this matches aoti but defers from the actual c10 DeviceGuard constructor that takes in device**)
    - `set_index(DeviceIndex)`
- `torch::stable::accelerator::Stream`: `std::shared_ptr` to `StreamOpaque`
     - constructor `Stream(StreamHandle stream)` (similar to torch::stable::Tensor)
     - `id() -> StreamId`
      
- `getCurrentStream(DeviceIndex device_index) -> stable::accelerator::Stream`


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #160453
* __->__ #159679

mikaylagawarecki committed:
Add beginning of torch::stable::accelerator

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add beginning of torch::stable::accelerator"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"




[ghstack-poisoned]

mikaylagawarecki commented on torch/csrc/stable/accelerator.h:
```diff
@@ -0,0 +1,68 @@
+
+using DeviceIndex = int8_t;
+using StreamId = int64_t;
+class DeviceGuard {
```
something that I'm not sure about -- do the copy / move semantics need to match the real device guard since this just a wrapper that is holding a `std::unique_ptr` to `DeviceGuardOpaque`?

janeyx99 replied:
It doesn't need to copy the existing DeviceGuard. For example, I agree with deleting the default constructor for this API. We can disallow copy and move if it's easiest to maintain anyway.

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"

Adds 
- `torch::stable::accelerator::DeviceGuard`: `std::unique_ptr` to `DeviceGuardOpauqe` copied from 

   https://github.com/pytorch/pytorch/blob/50eac811a68e63e96ad56c11c983bfe298a0bb8a/torch/csrc/inductor/aoti_runtime/utils_cuda.h#L30-L46
    - constructor `DeviceGuard(DeviceIndex)` (this matches aoti but defers from the actual c10 DeviceGuard constructor that takes in device
    - `set_index(DeviceIndex)`
- `torch::stable::accelerator::Stream`: `std::shared_ptr` to `StreamOpaque`
     - constructor `Stream(StreamHandle stream)` (similar to torch::stable::Tensor)
     - `id() -> StreamId`
      
- `getCurrentStream(DeviceIndex device_index) -> stable::accelerator::Stream`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"

Adds 
- `torch::stable::accelerator::DeviceGuard`: `std::unique_ptr` to `DeviceGuardOpauqe` copied from 

   https://github.com/pytorch/pytorch/blob/50eac811a68e63e96ad56c11c983bfe298a0bb8a/torch/csrc/inductor/aoti_runtime/utils_cuda.h#L30-L46
    - constructor `DeviceGuard(DeviceIndex)` (this matches aoti but defers from the actual c10 DeviceGuard constructor that takes in device
    - `set_index(DeviceIndex)`
- `torch::stable::accelerator::Stream`: `std::shared_ptr` to `StreamOpaque`
     - constructor `Stream(StreamHandle stream)` (similar to torch::stable::Tensor)
     - `id() -> StreamId`
      
- `getCurrentStream(DeviceIndex device_index) -> stable::accelerator::Stream`





[ghstack-poisoned]

albanD commented:
FYI @guangyey @EikanWang in case you have some feedback here.

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"

Adds 
- `torch::stable::accelerator::DeviceGuard`: `std::unique_ptr` to `DeviceGuardOpauqe` copied from 

   https://github.com/pytorch/pytorch/blob/50eac811a68e63e96ad56c11c983bfe298a0bb8a/torch/csrc/inductor/aoti_runtime/utils_cuda.h#L30-L46
    - constructor `DeviceGuard(DeviceIndex)` (this matches aoti but defers from the actual c10 DeviceGuard constructor that takes in device
    - `set_index(DeviceIndex)`
- `torch::stable::accelerator::Stream`: `std::shared_ptr` to `StreamOpaque`
     - constructor `Stream(StreamHandle stream)` (similar to torch::stable::Tensor)
     - `id() -> StreamId`
      
- `getCurrentStream(DeviceIndex device_index) -> stable::accelerator::Stream`





[ghstack-poisoned]

guangyey reviewed (approved):
Thanks.

janeyx99 commented on torch/csrc/stable/accelerator.h:
```diff
@@ -0,0 +1,69 @@
+
+  // Construct a stable::Stream from a StreamHandle
+  // Steals ownership from the StreamHandle
+  explicit Stream(StreamHandle stream)
```
Ah, I see, so the expected use case is:

- user calls getCurrentStream which creates a Stream
- then they can call id on it

janeyx99 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1590,3 +1594,53 @@ AOTITorchError aoti_torch_call_dispatcher(
+    StreamHandle* ret_stream) {
+  AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
+    c10::Stream stream = at::accelerator::getCurrentStream(device_index);
+    c10::Stream* stream_ptr = new c10::Stream(stream);
+    *ret_stream = reinterpret_cast<StreamHandle>(stream_ptr);
```
Ahhh after reading the rest of the code, I see what the user flow is intended to be. That said, I'm not sure if we want the semantics of get_current_stream to create a new stream on the heap (this is likely not expected from the user POV), and if we do end up sticking with this, we need to loudly document this as.

I think we do not want to mess with the memory of the stream at all...cc @albanD for thoughts

janeyx99 commented:
Thanks for taking this on! I'm left comments, i think my two biggest questions are:
1. How do we make sure DeviceGuard is working as expected haha
2. How do we want to represent Stream and what is the story of its memory handling

albanD commented on test/cpp_extensions/libtorch_agnostic_extension/setup.py:
```diff
@@ -35,10 +36,16 @@ def get_extension():
+    extension = CppExtension
+    # allow including <cuda_runtime.h>
+    if torch.cuda.is_available():
+        extra_compile_args["cxx"].append("-DUSE_CUDA")
```
Let's rename from USE_CUDA to LAE_USE_CUDA to avoid confusion with the USE_CUDA used in libtorch build

albanD commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/kernel.cpp:
```diff
@@ -320,3 +325,48 @@ STABLE_TORCH_LIBRARY_FRAGMENT(libtorch_agnostic, m) {
+  stack[0] = from(res);
+}
+
+int64_t test_stream(int8_t device_index) {
```
schema being int means the arg is int64_t

albanD commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/kernel.cpp:
```diff
@@ -320,3 +325,48 @@ STABLE_TORCH_LIBRARY_FRAGMENT(libtorch_agnostic, m) {
+// Test functions for torch::stable::accelerator APIs
+
+#ifdef USE_CUDA
+int test_device_guard(int8_t device_index) {
```
Plain "int" type should never be used. int64_t if you want a 64 bit int!

albanD commented on torch/csrc/stable/accelerator.h:
```diff
@@ -0,0 +1,69 @@
+
+} // namespace
+
+using DeviceIndex = int8_t; // this is from c10/core/Device.h
```
We should NOT rely on this one at the shim level.
For this part of the world, let's make it an int32_t for consistency with the existing shim.
We will have to change the tensor.h version.

albanD commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/kernel.cpp:
```diff
@@ -320,3 +325,48 @@ STABLE_TORCH_LIBRARY_FRAGMENT(libtorch_agnostic, m) {
+
+  DeviceGuard guard(device_index);
+  int currentDevice;
+  cudaError_t err = cudaGetDevice(&currentDevice);
```
Check that before the guard you have another device than this one.
Because if there is only 1 gpu, I'm not sure this will ever do anything

mikaylagawarecki commented on test/cpp_extensions/libtorch_agnostic_extension/setup.py:
```diff
@@ -35,10 +36,16 @@ def get_extension():
+    if torch.cuda.is_available():
+        extra_compile_args["cxx"].append("-DUSE_CUDA")
+        extension = CUDAExtension
+
```
Hmmm, I don't see the issues of the current approach, so gonna keep it as is unless there's something specific you're concerned about!

janeyx99 replied:
No issues yet, but if we have more "cuda only" code in the future it may be a good way to split the ever growing kernel.cpp

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"

Adds 
- `torch::stable::accelerator::DeviceGuard`: `std::unique_ptr` to `DeviceGuardOpauqe` mostly copied from the below (but made generic)

   https://github.com/pytorch/pytorch/blob/50eac811a68e63e96ad56c11c983bfe298a0bb8a/torch/csrc/inductor/aoti_runtime/utils_cuda.h#L30-L46
    - constructor `DeviceGuard(DeviceIndex)` (**this matches aoti but defers from the actual c10 DeviceGuard constructor that takes in device**)
    - `set_index(DeviceIndex)`
- `torch::stable::accelerator::Stream`: `std::shared_ptr` to `StreamOpaque`
     - constructor `Stream(StreamHandle stream)` (similar to torch::stable::Tensor)
     - `id() -> StreamId`
      
- `getCurrentStream(DeviceIndex device_index) -> stable::accelerator::Stream`





[ghstack-poisoned]

janeyx99 reviewed (approved):
lesgo!!

janeyx99 commented:
@pytorchbot merge -i

mikaylagawarecki commented on torch/csrc/stable/accelerator.h:
```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <torch/csrc/inductor/aoti_runtime/utils.h>
```
note to self to remove when rebasing and use `TORCH_ERROR_CODE_CHECK` instead

janeyx99 commented on torch/csrc/stable/accelerator.h:
```diff
@@ -0,0 +1,71 @@
+
+#include <torch/csrc/inductor/aoti_torch/c/shim.h>
+
+namespace torch::stable::accelerator {
```
nah let's follow the at:: example. ignore this!

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"

Adds 
- `torch::stable::accelerator::DeviceGuard`: `std::unique_ptr` to `DeviceGuardOpauqe` mostly copied from the below (but made generic)

   https://github.com/pytorch/pytorch/blob/50eac811a68e63e96ad56c11c983bfe298a0bb8a/torch/csrc/inductor/aoti_runtime/utils_cuda.h#L30-L46
    - constructor `DeviceGuard(DeviceIndex)` (**this matches aoti but defers from the actual c10 DeviceGuard constructor that takes in device**)
    - `set_index(DeviceIndex)`
- `torch::stable::accelerator::Stream`: `std::shared_ptr` to `StreamOpaque`
     - constructor `Stream(StreamHandle stream)` (similar to torch::stable::Tensor)
     - `id() -> StreamId`
      
- `getCurrentStream(DeviceIndex device_index) -> stable::accelerator::Stream`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"

Adds 
- `torch::stable::accelerator::DeviceGuard`: `std::unique_ptr` to `DeviceGuardOpauqe` mostly copied from the below (but made generic)

   https://github.com/pytorch/pytorch/blob/50eac811a68e63e96ad56c11c983bfe298a0bb8a/torch/csrc/inductor/aoti_runtime/utils_cuda.h#L30-L46
    - constructor `DeviceGuard(DeviceIndex)` (**this matches aoti but defers from the actual c10 DeviceGuard constructor that takes in device**)
    - `set_index(DeviceIndex)`
- `torch::stable::accelerator::Stream`: `std::shared_ptr` to `StreamOpaque`
     - constructor `Stream(StreamHandle stream)` (similar to torch::stable::Tensor)
     - `id() -> StreamId`
      
- `getCurrentStream(DeviceIndex device_index) -> stable::accelerator::Stream`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add beginnings of torch::stable::accelerator"

Adds 
- `torch::stable::accelerator::DeviceGuard`: `std::unique_ptr` to `DeviceGuardOpauqe` mostly copied from the below (but made generic)

   https://github.com/pytorch/pytorch/blob/50eac811a68e63e96ad56c11c983bfe298a0bb8a/torch/csrc/inductor/aoti_runtime/utils_cuda.h#L30-L46
    - constructor `DeviceGuard(DeviceIndex)` (**this matches aoti but defers from the actual c10 DeviceGuard constructor that takes in device**)
    - `set_index(DeviceIndex)`
- `torch::stable::accelerator::Stream`: `std::shared_ptr` to `StreamOpaque`
     - constructor `Stream(StreamHandle stream)` (similar to torch::stable::Tensor)
     - `id() -> StreamId`
      
- `getCurrentStream(DeviceIndex device_index) -> stable::accelerator::Stream`





[ghstack-poisoned]
